src/shm/pool/multi.rs

```diff
@@ -0,0 +1,159 @@
+//! A pool implementation which automatically frees buffers when released.
+
+use std::io;
+
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use wayland_client::ConnectionHandle;
+use wayland_client::protocol::{
+    wl_buffer,wl_shm
+};
+use wayland_client::{QueueHandle, Dispatch, DelegateDispatchBase, DelegateDispatch};
+
+use super::raw::RawPool;
+
+/// This pool manages multiple buffers associated with a surface.
+/// Only one buffer can be attributed to a surface.
+#[derive(Debug)]
+pub struct MultiPool<I: PartialEq + Clone> {
+    buffer_list: Vec<Buffer<I>>,
+    pub(crate) inner: RawPool,
+}
+
+#[derive(Debug)]
+struct Buffer<I: PartialEq + Clone> {
+    free: AtomicBool,
+    size: usize,
+    offset: usize,
+    buffer: wl_buffer::WlBuffer,
+    identifier: I,
+}
+
+impl<E: PartialEq + Clone> From<RawPool> for MultiPool<E> {
+    fn from(inner: RawPool) -> Self {
+        Self {
+            buffer_list: Vec::new(),
+            inner
+        }
+    }
+}
+
+impl<I: PartialEq + Clone> MultiPool<I> {
+    /// Resizes the memory pool, notifying the server the pool has changed in size.
+    ///
+    /// The wl_shm protocol only allows the pool to be made bigger. If the new size is smaller than the
+    /// current size of the pool, this function will do nothing.
+    pub fn resize(&mut self, size: usize, conn: &mut ConnectionHandle) -> io::Result<()> {
+        self.inner.resize(size, conn)
+    }
+    fn free(&self, buffer: &wl_buffer::WlBuffer) {
+        if let Some(buffer) = self.buffer_list.iter().find(|b| b.buffer.eq(buffer)) {
```

Comment:
.eq -> `==`

Reply:
since buffer is a reference as b.buffer isn't it's cleaner this way. the other option would be to clone buffer.

Reply:
I think `&b.buffer == buffer` is more direct for the eye (`==` is easily recognizable).